prize-tool/src/confirmation_latency.rs

```diff
@@ -0,0 +1,149 @@
+            voter_entry.last_slot = vote_state.votes.back().unwrap().slot;
+        }
+    }
+    slot_voters
+}
+
+// Assign latency scores to voters depending on how early their vote was recorded.
+fn score_voters(voters: &[HashSet<Pubkey>], voter_record: &mut HashMap<Pubkey, VoterEntry>) {
+    let total_voters: usize = voters.iter().map(|set| set.len()).sum();
+    let mut voters_seen = 0;
+    for voter_set in voters {
+        let is_low_latency = voters_seen < total_voters / 2;
+        let score_differential = if is_low_latency { 1 } else { -1 };
+        for voter in voter_set {
+            if let Some(voter_entry) = voter_record.get_mut(&voter) {
+                voter_entry.latency_score += score_differential;
+            }
+        }
+        voters_seen += voter_set.len();
+    }
+}
+
+fn normalize_winners(winners: &[(Pubkey, f64)]) -> Vec<Winner> {
+    winners
+        .iter()
+        .map(|(key, latency_score)| (*key, format!("Latency score: {:.*}", 0, latency_score)))
+        .collect()
+}
+
+/// Snapshot of the voting record of a validator
+pub type VoterRecord = HashMap<Pubkey, VoterEntry>;
+
+/// Ordered record of votes for each slot
+pub type SlotVoterSegments = BTreeMap<u64, Vec<HashSet<Pubkey>>>;
+
+/// Track voter latency by checkpointing the voter record after each entry.
+pub fn on_entry(
+    bank: &Bank,
+    voter_record: &mut HashMap<Pubkey, VoterEntry>,
+    slot_voter_segments: &mut BTreeMap<u64, Vec<HashSet<Pubkey>>>,
+) {
+    let mut slot_voters = voter_checkpoint(&bank, voter_record);
+    for (slot, voters) in slot_voters.drain() {
+        let slot_entry = slot_voter_segments.entry(slot).or_insert_with(Vec::new);
+        slot_entry.push(voters);
+    }
+    let bank_slot = bank.slot();
+
+    // Clear `slot_voter_segments` map when slot votes are old enough
+    let old_slots: Vec<_> = slot_voter_segments
+        .iter()
+        .map(|(slot, _)| *slot)
+        .take_while(|slot| *slot < bank_slot.saturating_sub(MAX_VOTE_DELAY))
+        .collect();
+    for old_slot in old_slots {
+        let voter_segments = slot_voter_segments.remove(&old_slot).unwrap();
+        score_voters(&voter_segments, voter_record);
+    }
+}
+
+pub fn compute_winners(
+    bank: &Bank,
+    baseline_id: &Pubkey,
+    voter_record: &mut VoterRecord,
+    slot_voter_segments: &mut SlotVoterSegments,
+) -> Winners {
+    // Score the remaining segments leftover from entry processing
+    for (_, voter_segments) in slot_voter_segments.iter() {
+        score_voters(voter_segments, voter_record);
+    }
+
+    let mut validator_latency: HashMap<Pubkey, i64> = HashMap::new();
+    for (voter_key, (_stake, account)) in bank.vote_accounts() {
+        let vote_state = VoteState::from(&account).unwrap();
+        let voter_entry = voter_record.remove(&voter_key).unwrap();
+        // It's possible that there are multiple vote accounts attributed to a validator
+        //   so use the max score when duplicates are found
```

Comment:
This seems inconsistent with the handling in rewards_earned, where multiple vote account rewards are summed.
What are the circumstances in which we envision validators ending up with multiple vote accounts?

Reply:
Yeah the inconsistency is no good, thanks for calling it out. My understanding is that we don't envision it happening but we currently don't prevent it from happening so I opted to handle it just in case.

That being said, here are some scenarios:
1. A validator could restart their node with a different vote account
2. A validator could disable the check that the vote account `node_pubkey` matches their node's pubkey. That way you could have two validator's colluding to inflate their earnings.
3. A validator could modify their code to run duplicate votes across multiple accounts

I was originally only thinking of the first scenario but I think now that to protect against gaming the winner calculations, we shouldn't be cumulating rewards across vote accounts. What do you think about using `max` for both "Confirmation Latency" and for "Highest Rewards Earned"?

Reply:
I like using `max` for both for now.
I'll think on it some more, though. Scenario 1 isn't really malicious behavior, and seems relatively likely to happen.

Reply:
Yeah, seems like a reasonable mistake to make. Sounds good 👍